### PR TITLE
add srcset transformer

### PR DESCRIPTION
**Pull request by sebastianbenz**

### CODE REVIEW

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+ *    pointing to a version of the `src` image with the given `width`. If no image is available, it should
+ *    return a falsy value. For example:
+ *
+ *       (src, width) => `${src}?width=${width}`
+ */
+class OptimizeImages {
+  constructor(config) {
+    this.log = config.log;
+    this.enabled = !!config.optimizeImages;
+    this.imageOptimizer = config.imageOptimizer || DEFAULT_IMAGE_OPTIMIZER;
```
Comment (sebastianbenz) [inline packages/optimizer/lib/transformers/OptimizeImages.js:158]: I've also considered to always require users to provide a function and get rid of the optimizeImages flag. Not sure if the default is very useful.

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+'use strict';
+
+const {hasAttribute, nextNode, firstChildByTag} = require('../NodeUtils');
+const {skipNodeAndChildren} = require('../HtmlDomHelper');
+const {isValidImageSrcURL} = require('../URLUtils');
+
+const MAX_IMG_SIZE = 820;
+const MIN_WIDTH_TO_ADD_SRCSET_IN_RESPONSIVE_LAYOUT = 300;
+const NUM_SRCSET_DPR = [1.0, 2.0, 3.0];
+const SRCSET_WIDTH = [
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:26]: whole lot of magic numbers, does it make sense to add a comment as to why these numbers are being used?

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+/**
+ * Default implementation that does not perform any kind of image optimization, but generates a `srcset` string
+ * by appending the width to the file name (e.g. image.jpg => image.32w.jpg).
+ *
+ * @param {string} src - the image src
+ * @param {number} width - the required widths (in px)
+ * @returns {string|undefined} - the image URL or undefined if no image is available in this dimension
+
+ */
+const DEFAULT_IMAGE_OPTIMIZER = (src, width) => {
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:57]: is this private? theres a lot of assumptions below (src exists, and is a string, with exists, etc)

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+ */
+class SrcsetWidth {
+  constructor() {
+    this.widthList_ = [];
+  }
+
+  /**
+   * Sets the base width, i.e., renderered dimension measured in CSS pixels.
+   * Returns true if srcset is needed, that is, we'll resize the image to at
+   * least 2 legitimate widths.
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:79]: what is a legitimate width?

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      if (width != previousWidth) {
+        this.widthList_.push(width);
+      }
+
+      previousWidth = width;
+    }
+  }
+
+  /**
+   *  Returns true if there is more legitimate width.
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:110]: im not sure if this means "returns true is there are additional widths that are also legitimate", or "returns true is there is a different width that is preferable to the current width"

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    return this.widthList_.length > 0;
+  }
+
+  /**
+   * Returns the current legitimate width and moves the state to the next one.
+   */
+  nextWidth() {
+    const nextWidth = this.widthList_[this.widthList_.length - 1];
+    this.widthList_.pop();
+    return nextWidth;
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:122]: any reason this wouldn't just be something like
```js
return this.widthList_.pop()
```

or is the variable allocation for readability?

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 1 above)
Comment (patrickkettner) [reply packages/optimizer/lib/transformers/OptimizeImages.js:158]: why is that?
My preference would be to either keep it or have a few examples documented somewheres to ease the onboarding

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+      return;
+    }
+    const src = imageNode.attribs.src;
+    // Check if it's a relative path or a valid http(s) URL.
+    if (!isValidImageSrcURL(src)) {
+      return;
+    }
+
+    // No srcset is added if the image ends with a `,` (comma). See
+    // http://b/127535381 for context.
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:198]: is it ok reference corplinks on a public repo?

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    if (src.endsWith(',')) {
+      return;
+    }
+    const width = imageNode.attribs.width;
+
+    // TODO(b/113271759): Handle width values that include 'px' (probably others).
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:204]: same as above

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 1 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:158]: My plan was to go with samples as well. Current thinking is to provide implementations for:

* cloudiary
* thumbor
* static site (via eleventy-img wrapper)
* custom sample

If we do this, then it probably makes sense to remove the default.

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 3 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:57]: will turn this into a sample, it's not useful this way

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 7 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:198]: yeah, it's not nice, but better than leaving it completely undocumented. It's a common pattern in the AMP repo.

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 2 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:26]: added comments

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 4 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:79]: added comment to the list of supported srcset widths

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 6 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:122]: no reason...

**Comment by sebastianbenz** [comment]
@patrickkettner PTAL

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 1 above)
Comment (patrickkettner) [reply packages/optimizer/lib/transformers/OptimizeImages.js]: what would downstream tools (WP, next, etc) do? provide their own functions?

Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js]: Correct, this is most likely something that users need to provide as frameworks usually don't optimize images by default.

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 5 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js]: rephrased

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+ * limitations under the License.
+ */
+
+'use strict';
+
+const {hasAttribute, nextNode, firstChildByTag} = require('../NodeUtils');
+const {skipNodeAndChildren} = require('../HtmlDomHelper');
+const {isValidImageSrcURL} = require('../URLUtils');
+
+// Don't generate srcsets for images larger than MAX_IMG_SIZE
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:23]: any reason this is the max that is worth documenting? based on stats? device? processing power?

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+const MAX_IMG_SIZE = 1500;
+
+// Don't generate srcset's for images with width smaller than MIN_WIDTH_TO_ADD_SRCSET_IN_RESPONSIVE_LAYOUT
+// this avoids generating srcsets for images with a responsive layout where width/height define the aspect ration.
+const MIN_WIDTH_TO_ADD_SRCSET_IN_RESPONSIVE_LAYOUT = 100;
+
+// All legimate srcset widths.
+const SRCSET_WIDTH = [39, 56, 82, 100, 150, 300, 500, 750, 1000, 1500, 2000, 2500];
+
+// The maximum number of values. We'll take the initial image width and generate more width values by
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:33]: does "values" in this context mean "maximum number of [`srcset` sources]"?

File: packages/optimizer/lib/transformers/OptimizeImages.js [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+ * - `imageOptimizer`: a function for customizing the srcset generation. The function should return a URL
+ *    pointing to a version of the `src` image with the given `width`. If no image is available, it should
+ *    return a falsy value. For example: (src, width) => `${src}?width=${width}`.
+ */
+class OptimizeImages {
+  constructor(config) {
+    this.log = config.log;
+    this.imageOptimizer = config.imageOptimizer;
+    this.srcsetWidth = new SrcsetWidth();
+    // TODO turn these into options
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeImages.js:125]: always cool to link to a tracking bug

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 9 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:23]: Good point. Leftover from cache implementation. Changed it to the max supported value.

File: packages/optimizer/lib/transformers/OptimizeImages.js (hunk 10 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeImages.js:33]: yes, rephrased

**Comment by sebastianbenz** [comment]
Thanks a lot for the great review @patrickkettner !

**Pull request merged**